### PR TITLE
[DOCS] Add `NoRepeatNGramLogitsProcessor` Example for `LogitsProcessor` class

### PR DESCRIPTION
**Pull request by Rishab26**
# What does this PR do?

This PR adds an example to the docstring of `NoRepeatNGramLogitsProcessor` and edits the docstring description of the same.
This is with reference to #24783 and is part of the #24575 

Fixes # (issue)
#24783

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).

## Who can review?

@gante @sgugger 

### Additional Notes:

In comparison with [TFNoRepeatNGramLogitsProcessor](https://github.com/huggingface/transformers/blob/05cda5df3405e6a2ee4ecf8f7e1b2300ebda472e/src/transformers/generation/tf_logits_process.py#L388), I noticed that the required functions `_get_ngrams`, `_get_generated_ngrams`, and `_calc_banned_ngram_tokens` were outside the class. Is this expected?

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -474,12 +476,43 @@ def _calc_banned_ngram_tokens(
 
 class NoRepeatNGramLogitsProcessor(LogitsProcessor):
     r"""
-    [`LogitsProcessor`] that enforces no repetition of n-grams. See
+    N-grams are groups of "n" consecutive words, characters, or tokens taken from a sequence of text. Given a sentence:
+    " She runs fast ", the bi-grams (n = 2) would be ("she","runs") and ("runs","fast"). In text generation, avoiding
+    repetitions of word sequences provides a more diverse and coherent output. This [`LogitsProcessor`] enforces no
```
Comment (gante) [inline src/transformers/generation/logits_process.py:481]:
```suggestion
    repetitions of word sequences provides a more diverse output. This [`LogitsProcessor`] enforces no
```
diverse -- absolutely
coherent -- depends on context, as you mention in your New York example below :)
(as such, removed the coherent)

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -474,12 +476,43 @@ def _calc_banned_ngram_tokens(
...
+    repetition of n-grams. See
     [Fairseq](https://github.com/pytorch/fairseq/blob/a07cb6f40480928c9e0548b737aadd36ee66ac76/fairseq/sequence_generator.py#L345).
+    It calls the <_calc_banned_ngram_tokens> function to determine the banned tokens (n-grams) based on the current
+    state of the generated sequences and the specified ngram_size. For each hypothesis, it sets the scores of banned
+    tokens to negative infinity (-inf), effectively discouraging the model from generating those tokens.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:486]:
```suggestion
```
These are implementation details, the user reading the docs doesn't need to know this. Shorter docs = user can find they need faster :)

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -474,12 +476,43 @@ def _calc_banned_ngram_tokens(
...
+
+    <Note> Use n-gram penalties with care. For instance, penalizing 2-grams (bigrams) in an article about the city of
+    New York might lead to undesirable outcomes where the city's name appears only once in the entire text.
+    [Reference](https://huggingface.co/blog/how-to-generate)
```
Comment (gante) [inline src/transformers/generation/logits_process.py:490]: This should be a `<Tip>` block. See [here](https://github.com/huggingface/transformers/blob/4a564490e149620282b9e946027e851f5f93aeae/src/transformers/generation/logits_process.py#L178) for an example

**Review by gante — COMMENTED** [review]
Thank you for opening the PR @Rishab26 🤗 I've added a few comments to make it more in line with our documentation :)

File: src/transformers/generation/logits_process.py (hunk 1 above)
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:481]: Makes sense. I will make the change.

File: src/transformers/generation/logits_process.py (hunk 2 above)
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:486]: Got it. 👍

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -462,9 +463,10 @@ def _calc_banned_ngram_tokens(
     if cur_len + 1 < ngram_size:
         # return no banned tokens if we haven't generated no_repeat_ngram_size tokens yet
         return [[] for _ in range(num_hypos)]
-
+    # With an example of no_repeat_ngram =2, generated ngrams look like this {(40,): [2883, 1053], (2883,): [2712], (2712,): [4346]}.
+    # The key is a generated word and the value is a list of the next words observed.
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:467]: Alright 👍

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -462,9 +463,10 @@ def _calc_banned_ngram_tokens(
...
     generated_ngrams = _get_ngrams(ngram_size, prev_input_ids, num_hypos)
-
+    # list of tokens that are banned, eg: [[], [836], [], [], [4346]]
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:469]: 👍

**Comment by Rishab26** [comment]
Hey @gante Thanks for your feedback. I've made the requested changes. Let me know what you think.

File: src/transformers/generation/logits_process.py [hunk 6]
```diff
@@ -438,6 +438,22 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 
 def _get_ngrams(ngram_size: int, prev_input_ids: torch.Tensor, num_hypos: int):
+    """
+    Parameters:
+    The key is a generated word and the value is a list of the next words observed.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:443]: This line is not part of the parameters, but part of the return type :)

File: src/transformers/generation/logits_process.py [hunk 7]
```diff
@@ -438,6 +438,22 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 
 def _get_ngrams(ngram_size: int, prev_input_ids: torch.Tensor, num_hypos: int):
+    """
+    Parameters:
```
Comment (gante) [inline src/transformers/generation/logits_process.py:442]:
```suggestion
    Args:
```

File: src/transformers/generation/logits_process.py [hunk 8]
```diff
@@ -438,6 +438,22 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+            `ngram_size` that can only occur once.
+        prev_input_ids (`torch.Tensor`):
+            A tensor containing tokenized input for each hypothesis.
+        num_hypos (`int`):
+            The number of hypotheses for which n-grams need to be generated.
+
+    Returns:
+        Dictionary of generated ngrams. With an example of no_repeat_ngram =2, assuming input tokenized tensor([[ 40,
+        2883, 2712, 4346, 319, 262, 21511]]) generated ngrams look like this {(40,): [2883, 1053], (2883,): [2712],
+        (2712,): [4346]}.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:454]: I would either:
a) remove the example
b) move the example to before `Args`, in a sentence explaining what the function does and showing an example

File: src/transformers/generation/logits_process.py [hunk 9]
```diff
@@ -450,6 +466,21 @@ def _get_ngrams(ngram_size: int, prev_input_ids: torch.Tensor, num_hypos: int):
 
 
 def _get_generated_ngrams(banned_ngrams, prev_input_ids, ngram_size, cur_len):
+    """
+    Parameters:
```
Comment (gante) [inline src/transformers/generation/logits_process.py:470]:
```suggestion
    Args:
```

File: src/transformers/generation/logits_process.py [hunk 10]
```diff
@@ -450,6 +466,21 @@ def _get_ngrams(ngram_size: int, prev_input_ids: torch.Tensor, num_hypos: int):
...
+            A dictionary containing previously generated n-grams for each hypothesis.
+        prev_input_ids (`torch.Tensor`):
+            A `tensor` containing tokenized input for each hypothesis in the current batch.
+        ngram_size (int):
+            `ngram_size` that can only occur once.
+        cur_len (`int`):
+            The current length of the token sequences for which the n-grams are being checked.
+
+    Returns:
+        List of tokens that are banned, output eg: [[], [836], [], [], [4346]]
```
Comment (gante) [inline src/transformers/generation/logits_process.py:481]: Same example-related comment :)

**Review by gante — COMMENTED** [review]
Thanks for iterating! A few docstring-related nits, to make it uniform with the rest of the library :)

File: src/transformers/generation/logits_process.py (hunk 6 above)
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:443]: Oops! Thanks :)

File: src/transformers/generation/logits_process.py (hunk 7 above)
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:442]: 👍

File: src/transformers/generation/logits_process.py (hunk 9 above)
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:470]: 👍

File: src/transformers/generation/logits_process.py (hunk 8 above)
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:454]: Let me try again :)

File: src/transformers/generation/logits_process.py (hunk 10 above)
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:481]: Yep, on it.

**Comment by Rishab26** [comment]
@gante I've made some changes. Does this look better?

**Review by gante — APPROVED** [review]
LGTM, thanks for iterating 👍

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Most of the comments are about formatting. The main one is making sure that the information in the docstrings are correct and terse.

File: src/transformers/generation/logits_process.py [hunk 11]
```diff
@@ -438,17 +438,59 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 
 def _get_ngrams(ngram_size: int, prev_input_ids: torch.Tensor, num_hypos: int):
+    """
+    Assume ngram_size =2 and prev_input_ids= tensor([[ 40, 2883, 2712, 4346, 319, 262, 21511]]). The output of
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:442]: 👍

File: src/transformers/generation/logits_process.py [hunk 12]
```diff
@@ -438,17 +438,59 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+    generated ngrams look like this {(40,): [2883, 1053], (2883,): [2712], (2712,): [4346]} (Note this is only a subset
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:443]: It's a good point. I did not want to place the whole tensor here with 1053, so I added a note stating this is only a subset and the dictionary was to illustrate a simplified representation of the generated n-gram. 

However, I can see how this won't be very clear. I will attempt to abstract this without values to make it easier to understand.

File: src/transformers/generation/logits_process.py [hunk 13]
```diff
@@ -438,17 +438,59 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
 
 
 def _get_generated_ngrams(banned_ngrams, prev_input_ids, ngram_size, cur_len):
+    """
+    This function is responsible for determining the tokens that are banned for each hypothesis based on the previously
+    generated n-grams. Let's consider generated_ngrams = { (40,): [2883, 1053], (2883,): [2712], (2712,): [4346] },
+    prev_input_ids = torch.tensor([[40, 2883, 2712, 4346, 319, 262, 21511]]), cur_len = 3. The function then calculates
+    the banned tokens for each index by looking at the last ngram_size - 1 tokens (excluding the last token) of the
+    current hypothesis and checking if they form an n-gram that exists in the banned_ngrams dictionary. In this case,
+    start_idx = 2 ngram_idx = (2712,) which matches the key from the banned_ngrams list output returns: [4346]
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:478]: > We don't really need to introduce it's a function. A short, single-line description first is best.
I think it is best to leave it at that. I'll make the change 👍 Let me know what you think.

File: src/transformers/generation/logits_process.py [hunk 14]
```diff
@@ -438,17 +438,59 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    Args:
+        banned_ngrams (`dict`):
+            A dictionary containing previously generated n-grams for each hypothesis.
+        prev_input_ids (`torch.Tensor`):
+            A `tensor` containing tokenized input for each hypothesis in the current batch.
+        ngram_size (int):
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:485]: 👍

File: src/transformers/generation/logits_process.py [hunk 15]
```diff
@@ -474,12 +514,43 @@ def _calc_banned_ngram_tokens(
 
 class NoRepeatNGramLogitsProcessor(LogitsProcessor):
     r"""
-    [`LogitsProcessor`] that enforces no repetition of n-grams. See
+    N-grams are groups of "n" consecutive words, characters, or tokens taken from a sequence of text. Given a sentence:
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:517]: yes :)

File: src/transformers/generation/logits_process.py [hunk 16]
```diff
@@ -474,12 +514,43 @@ def _calc_banned_ngram_tokens(
...
+    " She runs fast ", the bi-grams (n = 2) would be ("she","runs") and ("runs","fast"). In text generation, avoiding
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:518]: thank you 👍

**Comment by Rishab26** [comment]
@amyeroberts I've made some changes. Let me know what you think.

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating! 

Just a few small notes on the correctness of the docstrings / styling nits. 

Note: What's being used in these functions are token ids not tokens. When a sentence is tokenized, it typically means splitting the words into word pieces e.g. `"This is a tokenized sentence"` -> `['this', 'is', 'a', 'token', '##ized', 'sentence']` i.e. it is a list of strings. These are then mapped to ints, which are token ids and that we pass to the model. 

We're not completely strict everywhere with using token ids vs tokens and someone will understand if you say `list of generated tokens`. However `tokenized list of next words observed` is wrong - the list isn't tokenized and it's not next words but ints that map to word-pieces. We should say something like `list of next token ids`.

**Comment by Rishab26** [comment]
> Thanks for iterating!
> 
> Just a few small notes on the correctness of the docstrings / styling nits.
> 
> Note: What's being used in these functions are token ids not tokens. When a sentence is tokenized, it typically means splitting the words into word pieces e.g. `"This is a tokenized sentence"` -> `['this', 'is', 'a', 'token', '##ized', 'sentence']` i.e. it is a list of strings. These are then mapped to ints, which are token ids and that we pass to the model.
> 
> We're not completely strict everywhere with using token ids vs tokens and someone will understand if you say `list of generated tokens`. However `tokenized list of next words observed` is wrong - the list isn't tokenized and it's not next words but ints that map to word-pieces. We should say something like `list of next token ids`.

Of course, I understand that :) I was only trying to make that function simpler to understand. Having said that, you're absolutely right,  `list of next token ids` is the right framing. I'll make the change.

File: src/transformers/generation/logits_process.py [hunk 17]
```diff
@@ -438,17 +438,53 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
 
 def _get_ngrams(ngram_size: int, prev_input_ids: torch.Tensor, num_hypos: int):
+    """
+    Assume ngram_size=2 and prev_input_ids=tensor([[list of generated tokens]]). The output of generated ngrams look
+    like this {(generated word #1,): [tokenized list of next words observed], (generated word #2,): [tokenized list of
+    next words observed] }.
+
+    Args:
+        ngram_size (`int`):
+            `ngram_size` that can only occur once.
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:448]: Okay 👍

File: src/transformers/generation/logits_process.py [hunk 18]
```diff
@@ -438,17 +438,53 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 
 def _get_ngrams(ngram_size: int, prev_input_ids: torch.Tensor, num_hypos: int):
+    """
+    Assume ngram_size=2 and prev_input_ids=tensor([[list of generated tokens]]). The output of generated ngrams look
+    like this {(generated word #1,): [tokenized list of next words observed], (generated word #2,): [tokenized list of
+    next words observed] }.
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:444]: Let me try again :)

File: src/transformers/generation/logits_process.py [hunk 19]
```diff
@@ -438,17 +438,53 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+    """
+    Determines the banned tokens for each hypothesis based on previously generated n-grams.
+
+    Args:
+        banned_ngrams (`dict`):
+            A dictionary containing previously generated n-grams for each hypothesis.
+        prev_input_ids (`torch.Tensor`):
+            A `tensor` containing tokenized input for each hypothesis in the current batch.
+        ngram_size (`int`):
+            `ngram_size` that can only occur once.
```
[the remark this answers is not in the document]
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:480]: 👍

**Comment by Rishab26** [comment]
@amyeroberts Thanks for your feedback. Let me know if the changes look good now 👍

File: src/transformers/generation/logits_process.py [hunk 20]
```diff
@@ -474,12 +507,43 @@ def _calc_banned_ngram_tokens(
 
 class NoRepeatNGramLogitsProcessor(LogitsProcessor):
     r"""
-    [`LogitsProcessor`] that enforces no repetition of n-grams. See
+    N-grams are groups of "n" consecutive words, characters, or tokens taken from a sequence of text. Given the
+    sentence: "She runs fast", the bi-grams (n = 2) would be ("she","runs") and ("runs","fast"). In text generation,
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:511]: nit: pep8

```suggestion
    sentence: "She runs fast", the bi-grams (n=2) would be ("she", "runs") and ("runs", "fast"). In text generation,
```

File: src/transformers/generation/logits_process.py [hunk 21]
```diff
@@ -438,17 +438,52 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+        prev_input_ids (`torch.Tensor`):
+            Generated token ids for the current hypothesis.
+        ngram_size (`int`):
+            The number sequential tokens taken as a group which may only occur once before being banned.
+        cur_len (`int`):
+            The current length of the token sequences for which the n-grams are being checked.
+
+    Returns:
+        List of tokens that are banned.
+
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:485]:
```suggestion
```

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating! 

Just a tiny tiny nit, and I think we're good to merge!

File: src/transformers/generation/logits_process.py (hunk 20 above)
Comment (Rishab26) [reply src/transformers/generation/logits_process.py:511]: I missed this, sorry you had to point this out twice. It's a good way to internalize this. Making this change right away.

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating!
